src/btree/bt_handle.c

```diff
@@ -854,12 +854,12 @@ __btree_get_last_recno(WT_SESSION_IMPL *session)
     }
 
     /*
-     * The endpoint for append is global; read the last page with global visibility to make sure
-     * that if the end of the tree is truncated we don't start appending in the truncated space
-     * unless the truncation has become globally visible. (Note that this path does not examine the
-     * visibility of individual data items; it only checks whether whole pages are deleted.)
+     * The endpoint for append is global; read the last page with global visibility (even if it's
+     * deleted) to make sure that if the end of the tree is truncated, the tree walk finds the
+     * correct page. (Note that this path does not examine the visibility of individual data items;
+     * it only checks whether whole pages are deleted.)
      */
-    flags = WT_READ_PREV | WT_READ_VISIBLE_ALL;
+    flags = WT_READ_PREV | WT_READ_VISIBLE_ALL | WT_READ_SEE_DELETED;
```

Comment:
As we are skipping the deleted pages, the flag `WT_READ_VISIBLE_ALL` is no longer needed. The flag usage was to skip the globally visible deleted pages and not just visible deleted pages. Also, the comment can be updated to remove the `global visibility`.

Reply:
Hari and I discussed this offline. There's a few things going on with this flag:
* It's currently only used to skip globally visible deletes. Given this, the name of `WT_READ_VISIBLE_ALL` is very confusing.
* There are two problems with renaming the flag: (1) it might reasonably be extended to cover all global visibility, and (2) it means the API would now have `WT_READ_SEE_DELETED` and `WT_READ_SKIP_GLOBALLY_VISIBLE_DELETES` (or whatever the name would be). This is very confusing and the interaction isn't obvious at all.
* `__tree_walk_internal` is almost using it for global visibility in general already, given this assert:
```
/* Ensure we have a snapshot to check visibility or we only check global visibility. */
WT_ASSERT(session, LF_ISSET(WT_READ_VISIBLE_ALL) || F_ISSET(session->txn, WT_TXN_HAS_SNAPSHOT));
```

So our conclusion was that this `flags` variable is fine as-is.

Reply:
I think I understand what is happening with the flags here. I'd like a follow up ticket created to tidy up the flags after this PR is merged.

Reply:
Made WT-11683 for the flags tidy-up.